File: src/opochhash/poc/field.py
"""
Goldilocks Prime Field Arithmetic

Field: F_p where p = 2^64 - 2^32 + 1

This prime has special structure enabling:
- Fast modular reduction
- FFT-friendly (has large 2-adic subgroup)
- Efficient SIMD implementation

The multiplicative group has order p-1 = 2^32 * (2^32 - 1)
which contains a subgroup of order 2^32 (for FFT up to 2^32 elements).
"""

from __future__ import annotations
from typing import List, Tuple, Optional
import random


# Goldilocks prime: p = 2^64 - 2^32 + 1
GOLDILOCKS_PRIME = (1 << 64) - (1 << 32) + 1

# Generator of multiplicative group (primitive root)
MULTIPLICATIVE_GENERATOR = 7

# Two-adicity: largest k such that 2^k divides p-1
TWO_ADICITY = 32

# 2^32-th root of unity
ROOT_OF_UNITY_ORDER_2_32 = pow(MULTIPLICATIVE_GENERATOR, (GOLDILOCKS_PRIME - 1) >> TWO_ADICITY, GOLDILOCKS_PRIME)


class FieldElement:
    """
    Element of the Goldilocks prime field.

    Represents values in F_p where p = 2^64 - 2^32 + 1.

    All arithmetic operations are constant-time modular operations.
    """

    __slots__ = ('value',)

    def __init__(self, value: int):
        """Create field element from integer."""
        self.value = value % GOLDILOCKS_PRIME

    # =========================================================================
    # Arithmetic Operations
    # =========================================================================

    def __add__(self, other: FieldElement) -> FieldElement:
        """Addition in F_p."""
        return FieldElement(self.value + other.value)

    def __sub__(self, other: FieldElement) -> FieldElement:
        """Subtraction in F_p."""
        return FieldElement(self.value - other.value + GOLDILOCKS_PRIME)

    def __mul__(self, other: FieldElement) -> FieldElement:
        """Multiplication in F_p."""
        return FieldElement(self.value * other.value)

    def __neg__(self) -> FieldElement:
        """Negation in F_p."""
        return FieldElement(GOLDILOCKS_PRIME - self.value if self.value else 0)

    def __truediv__(self, other: FieldElement) -> FieldElement:
        """Division in F_p (multiplication by inverse)."""
        return self * other.inverse()

    def __pow__(self, exp: int) -> FieldElement:
        """Exponentiation using square-and-multiply."""
        if exp < 0:
            return self.inverse() ** (-exp)
        return FieldElement(pow(self.value, exp, GOLDILOCKS_PRIME))

    def inverse(self) -> FieldElement:
        """
        Multiplicative inverse using Fermat's little theorem.

        a^-1 = a^(p-2) mod p
        """
        if self.value == 0:
            raise ZeroDivisionError("Cannot invert zero")
        return FieldElement(pow(self.value, GOLDILOCKS_PRIME - 2, GOLDILOCKS_PRIME))

    def sqrt(self) -> Optional[FieldElement]:
        """
        Square root if it exists.

        For Goldilocks, p ≡ 1 (mod 4), so we use Tonelli-Shanks.
        """
        return _tonelli_shanks(self)

    # =========================================================================
    # Comparison Operations
    # =========================================================================

    def __eq__(self, other: object) -> bool:
        if isinstance(other, FieldElement):
            return self.value == other.value
        if isinstance(other, int):
            return self.value == (other % GOLDILOCKS_PRIME)
        return False

    def __ne__(self, other: object) -> bool:
        return not self.__eq__(other)

    def __hash__(self) -> int:
        return hash(self.value)

    def __repr__(self) -> str:
        return f"FieldElement({self.value})"

    def __str__(self) -> str:
        return str(self.value)

    # =========================================================================
    # Serialization
    # =========================================================================

    def to_bytes(self) -> bytes:
        """Serialize to 8 bytes (little-endian)."""
        return self.value.to_bytes(8, 'little')

    @classmethod
    def from_bytes(cls, data: bytes) -> FieldElement:
        """Deserialize from 8 bytes (little-endian)."""
        return cls(int.from_bytes(data[:8], 'little'))

    def to_int(self) -> int:
        """Convert to integer."""
        return self.value

    # =========================================================================
    # Predicates
    # =========================================================================

    def is_zero(self) -> bool:
        return self.value == 0

    def is_one(self) -> bool:
        return self.value == 1

    # =========================================================================
    # Class Methods
    # =========================================================================

    @classmethod
    def zero(cls) -> FieldElement:
        """Additive identity."""
        return cls(0)

    @classmethod
    def one(cls) -> FieldElement:
        """Multiplicative identity."""
        return cls(1)

    @classmethod
    def random(cls) -> FieldElement:
        """Generate random field element."""
        return cls(random.randrange(GOLDILOCKS_PRIME))

    @classmethod
    def from_hash(cls, data: bytes) -> FieldElement:
        """Create field element from hash output (reduces mod p)."""
        # Take first 8 bytes, reduce mod p
        value = int.from_bytes(data[:8], 'little')
        return cls(value)

    @classmethod
    def primitive_root_of_unity(cls, order: int) -> FieldElement:
        """
        Get primitive n-th root of unity where n = 2^order.

        Requires order <= TWO_ADICITY (32).
        """
        if order > TWO_ADICITY:
            raise ValueError(f"Order 2^{order} exceeds two-adicity {TWO_ADICITY}")

        # Start with 2^32-th root of unity and square down
        root = cls(ROOT_OF_UNITY_ORDER_2_32)
        for _ in range(TWO_ADICITY - order):
            root = root * root

        return root


# =============================================================================
# Helper Functions
# =============================================================================

def _tonelli_shanks(a: FieldElement) -> Optional[FieldElement]:
    """
    Tonelli-Shanks algorithm for computing square roots.

    Returns sqrt(a) if it exists, None otherwise.
    """
    if a.is_zero():
        return FieldElement.zero()

    p = GOLDILOCKS_PRIME

    # Check if a is a quadratic residue
    if pow(a.value, (p - 1) // 2, p) != 1:
        return None

    # Factor p-1 = 2^s * q where q is odd
    s = TWO_ADICITY
    q = (p - 1) >> s

    # Find a quadratic non-residue
    z = 2
    while pow(z, (p - 1) // 2, p) != p - 1:
        z += 1

    m = s
    c = pow(z, q, p)
    t = pow(a.value, q, p)
    r = pow(a.value, (q + 1) // 2, p)

    while True:
        if t == 1:
            return FieldElement(r)

        # Find least i such that t^(2^i) = 1
        i = 1
        temp = (t * t) % p
        while temp != 1:
            temp = (temp * temp) % p
            i += 1

        # Update
        b = pow(c, 1 << (m - i - 1), p)
        m = i
        c = (b * b) % p
        t = (t * c) % p
        r = (r * b) % p


def batch_inverse(elements: List[FieldElement]) -> List[FieldElement]:
    """
    Batch inversion using Montgomery's trick.

    Computes inverses of n elements using 3(n-1) multiplications + 1 inversion.
    Much faster than n individual inversions.
    """
    n = len(elements)
    if n == 0:
        return []
    if n == 1:
        return [elements[0].inverse()]

    # Forward pass: compute prefix products
    prefix = [FieldElement.one()] * n
    prefix[0] = elements[0]
    for i in range(1, n):
        prefix[i] = prefix[i-1] * elements[i]

    # Single inversion of the total product
    inv_total = prefix[-1].inverse()

    # Backward pass: compute individual inverses
    inverses = [FieldElement.zero()] * n
    for i in range(n - 1, 0, -1):
        inverses[i] = inv_total * prefix[i-1]
        inv_total = inv_total * elements[i]
    inverses[0] = inv_total

    return inverses


def interpolate_at_point(
    points: List[Tuple[FieldElement, FieldElement]],
    x: FieldElement
) -> FieldElement:
    """
    Evaluate interpolating polynomial at a point using Lagrange interpolation.

    Given points [(x_0, y_0), ..., (x_n, y_n)], compute P(x) where
    P is the unique polynomial of degree ≤ n passing through all points.
    """
    n = len(points)
    result = FieldElement.zero()

    for i in range(n):
        xi, yi = points[i]

        # Compute Lagrange basis polynomial L_i(x)
        numerator = FieldElement.one()
        denominator = FieldElement.one()

        for j in range(n):
            if i != j:
                xj, _ = points[j]
                numerator = numerator * (x - xj)
                denominator = denominator * (xi - xj)

        # L_i(x) = numerator / denominator
        basis = numerator / denominator

        # Add y_i * L_i(x) to result
        result = result + (yi * basis)

    return result


# =============================================================================
# Extension Field (for FRI if needed)
# =============================================================================

class QuadraticExtension:
    """
    Quadratic extension of Goldilocks: F_p[x] / (x^2 - 7)

    Elements are a + b*w where w^2 = 7 (non-residue).
    """

    __slots__ = ('a', 'b')

    # Non-residue for extension
    NON_RESIDUE = FieldElement(7)

    def __init__(self, a: FieldElement, b: FieldElement):
        self.a = a
        self.b = b

    def __add__(self, other: QuadraticExtension) -> QuadraticExtension:
        return QuadraticExtension(self.a + other.a, self.b + other.b)

    def __sub__(self, other: QuadraticExtension) -> QuadraticExtension:
        return QuadraticExtension(self.a - other.a, self.b - other.b)

    def __mul__(self, other: QuadraticExtension) -> QuadraticExtension:
        # (a + bw)(c + dw) = (ac + bd*7) + (ad + bc)w
        a, b = self.a, self.b
        c, d = other.a, other.b
        return QuadraticExtension(
            a * c + b * d * self.NON_RESIDUE,
            a * d + b * c
        )

    def __neg__(self) -> QuadraticExtension:
        return QuadraticExtension(-self.a, -self.b)

    def inverse(self) -> QuadraticExtension:
        # (a + bw)^-1 = (a - bw) / (a^2 - 7*b^2)
        denom = self.a * self.a - self.NON_RESIDUE * self.b * self.b
        denom_inv = denom.inverse()
        return QuadraticExtension(self.a * denom_inv, -self.b * denom_inv)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, QuadraticExtension):
            return self.a == other.a and self.b == other.b
        return False

    @classmethod
    def zero(cls) -> QuadraticExtension:
        return cls(FieldElement.zero(), FieldElement.zero())

    @classmethod
    def one(cls) -> QuadraticExtension:
        return cls(FieldElement.one(), FieldElement.zero())

    def to_bytes(self) -> bytes:
        return self.a.to_bytes() + self.b.to_bytes()

    @classmethod
    def from_bytes(cls, data: bytes) -> QuadraticExtension:
        return cls(
            FieldElement.from_bytes(data[:8]),
            FieldElement.from_bytes(data[8:16])
        )

File: opoch-poc-sha/rust-verifier/src/field.rs
//! Goldilocks Prime Field Arithmetic
//!
//! p = 2^64 - 2^32 + 1 = 18446744069414584321
//!
//! This field has excellent properties:
//! - 64-bit operations fit in registers
//! - Two-adicity of 32 (supports FFT up to 2^32)
//! - Fast reduction using the special form

use std::ops::{Add, Sub, Mul, Div, Neg};

/// Goldilocks prime: p = 2^64 - 2^32 + 1
pub const GOLDILOCKS_PRIME: u64 = 0xFFFFFFFF00000001;

/// Two-adicity: largest k such that 2^k | (p-1)
pub const TWO_ADICITY: u32 = 32;

/// Generator of the multiplicative group
pub const GENERATOR: u64 = 7;

/// Primitive 2^32-th root of unity
/// This is GENERATOR^((p-1)/2^32) = 7^(2^32 - 1) mod p
pub const PRIMITIVE_ROOT_2_32: u64 = 1753635133440165772;

/// Field element in Goldilocks field
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Fp(pub u64);

impl Fp {
    /// Zero element
    pub const ZERO: Fp = Fp(0);

    /// One element
    pub const ONE: Fp = Fp(1);

    /// Create from u64, reducing mod p
    #[inline]
    pub fn new(x: u64) -> Self {
        Fp(Self::reduce(x as u128))
    }

    /// Create from u128, reducing mod p
    #[inline]
    pub fn from_u128(x: u128) -> Self {
        Fp(Self::reduce(x))
    }

    /// Reduce a u128 modulo p using the special form of Goldilocks
    /// p = 2^64 - 2^32 + 1, so 2^64 ≡ 2^32 - 1 (mod p)
    #[inline]
    fn reduce(x: u128) -> u64 {
        // For now, use the simple approach
        // This can be optimized later with the special form
        (x % (GOLDILOCKS_PRIME as u128)) as u64
    }

    /// Optimized reduction for values that fit in u128
    /// Uses the identity: 2^64 ≡ 2^32 - 1 (mod p)
    #[inline]
    #[allow(dead_code)]
    fn reduce_fast(x: u128) -> u64 {
        // Split x = high * 2^64 + low
        let low = x as u64;
        let high = (x >> 64) as u64;

        if high == 0 {
            return if low >= GOLDILOCKS_PRIME {
                low - GOLDILOCKS_PRIME
            } else {
                low
            };
        }

        // 2^64 ≡ 2^32 - 1 (mod p)
        // high * 2^64 ≡ high * 2^32 - high (mod p)

        // Compute high * 2^32 - this fits in u64 if high < 2^32
        // For high >= 2^32, we need to handle overflow
        let (high_times_2_32, overflow1) = (high as u128).overflowing_mul(1u128 << 32);

        // Compute low + high * 2^32 - high
        let sum = (low as u128)
            .wrapping_add(high_times_2_32 as u128)
            .wrapping_sub(high as u128);

        // Handle any new overflow from the sum
        let result_low = sum as u64;
        let result_high = (sum >> 64) as u64;

        if result_high == 0 {
            if result_low >= GOLDILOCKS_PRIME {
                result_low - GOLDILOCKS_PRIME
            } else {
                result_low
            }
        } else {
            // Recurse (should only happen once)
            Self::reduce(sum)
        }
    }

    /// Check if zero
    #[inline]
    pub fn is_zero(&self) -> bool {
        self.0 == 0
    }

    /// Multiplicative inverse using Fermat's little theorem
    /// a^(-1) = a^(p-2) mod p
    pub fn inverse(&self) -> Self {
        if self.is_zero() {
            panic!("Cannot invert zero");
        }
        self.pow(GOLDILOCKS_PRIME - 2)
    }

    /// Exponentiation by squaring
    pub fn pow(&self, exp: u64) -> Self {
        let mut result = Fp::ONE;
        let mut base = *self;
        let mut e = exp;

        while e > 0 {
            if e & 1 == 1 {
                result = result * base;
            }
            base = base * base;
            e >>= 1;
        }

        result
    }

    /// Get primitive n-th root of unity (n must be power of 2, n <= 2^32)
    /// Returns ω such that ω^n = 1 and ω^k ≠ 1 for 0 < k < n
    pub fn primitive_root_of_unity(log_n: u32) -> Self {
        assert!(log_n <= TWO_ADICITY, "Root order too large");

        // PRIMITIVE_ROOT_2_32 is the primitive 2^32-th root of unity
        // To get 2^log_n-th root, we raise it to 2^(32 - log_n)
        // Because (ω^(2^(32-log_n)))^(2^log_n) = ω^(2^32) = 1
        let omega_2_32 = Fp::new(PRIMITIVE_ROOT_2_32);
        omega_2_32.pow(1u64 << (TWO_ADICITY - log_n))
    }

    /// Convert to bytes (little-endian)
    pub fn to_bytes(&self) -> [u8; 8] {
        self.0.to_le_bytes()
    }

    /// Convert from bytes (little-endian)
    pub fn from_bytes(bytes: &[u8]) -> Self {
        let mut arr = [0u8; 8];
        arr.copy_from_slice(&bytes[..8]);
        let x = u64::from_le_bytes(arr);
        Fp::new(x)
    }

    /// Get the raw u64 value
    pub fn to_u64(&self) -> u64 {
        self.0
    }
}

impl Add for Fp {
    type Output = Self;

    #[inline]
    fn add(self, rhs: Self) -> Self {
        let sum = self.0 as u128 + rhs.0 as u128;
        if sum >= GOLDILOCKS_PRIME as u128 {
            Fp((sum - GOLDILOCKS_PRIME as u128) as u64)
        } else {
            Fp(sum as u64)
        }
    }
}

impl Sub for Fp {
    type Output = Self;

    #[inline]
    fn sub(self, rhs: Self) -> Self {
        if self.0 >= rhs.0 {
            Fp(self.0 - rhs.0)
        } else {
            Fp(GOLDILOCKS_PRIME - rhs.0 + self.0)
        }
    }
}

impl Mul for Fp {
    type Output = Self;

    #[inline]
    fn mul(self, rhs: Self) -> Self {
        let prod = self.0 as u128 * rhs.0 as u128;
        Fp(Self::reduce(prod))
    }
}

impl Div for Fp {
    type Output = Self;

    fn div(self, rhs: Self) -> Self {
        self * rhs.inverse()
    }
}

impl Neg for Fp {
    type Output = Self;

    fn neg(self) -> Self {
        if self.0 == 0 {
            self
        } else {
            Fp(GOLDILOCKS_PRIME - self.0)
        }
    }
}

/// Quadratic extension field F_{p^2} = F_p[x] / (x^2 - 7)
/// Elements: a + b*α where α^2 = 7
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Fp2 {
    pub c0: Fp, // Real part
    pub c1: Fp, // Imaginary part (coefficient of α)
}

/// Non-residue: α^2 = 7
const NON_RESIDUE: u64 = 7;

impl Fp2 {
    pub const ZERO: Fp2 = Fp2 { c0: Fp::ZERO, c1: Fp::ZERO };
    pub const ONE: Fp2 = Fp2 { c0: Fp::ONE, c1: Fp::ZERO };

    pub fn new(c0: Fp, c1: Fp) -> Self {
        Fp2 { c0, c1 }
    }

    pub fn from_fp(x: Fp) -> Self {
        Fp2 { c0: x, c1: Fp::ZERO }
    }

    pub fn is_zero(&self) -> bool {
        self.c0.is_zero() && self.c1.is_zero()
    }

    /// Multiplicative inverse
    pub fn inverse(&self) -> Self {
        // (a + bα)^(-1) = (a - bα) / (a^2 - 7*b^2)
        let norm = self.c0 * self.c0 - Fp::new(NON_RESIDUE) * self.c1 * self.c1;
        let norm_inv = norm.inverse();
        Fp2 {
            c0: self.c0 * norm_inv,
            c1: -self.c1 * norm_inv,
        }
    }

    /// Exponentiation
    pub fn pow(&self, exp: u64) -> Self {
        let mut result = Fp2::ONE;
        let mut base = *self;
        let mut e = exp;

        while e > 0 {
            if e & 1 == 1 {
                result = result * base;
            }
            base = base * base;
            e >>= 1;
        }

        result
    }
}

impl Add for Fp2 {
    type Output = Self;

    fn add(self, rhs: Self) -> Self {
        Fp2 {
            c0: self.c0 + rhs.c0,
            c1: self.c1 + rhs.c1,
        }
    }
}

impl Sub for Fp2 {
    type Output = Self;

    fn sub(self, rhs: Self) -> Self {
        Fp2 {
            c0: self.c0 - rhs.c0,
            c1: self.c1 - rhs.c1,
        }
    }
}

impl Mul for Fp2 {
    type Output = Self;

    fn mul(self, rhs: Self) -> Self {
        // (a + bα)(c + dα) = (ac + 7bd) + (ad + bc)α
        let ac = self.c0 * rhs.c0;
        let bd = self.c1 * rhs.c1;
        let ad = self.c0 * rhs.c1;
        let bc = self.c1 * rhs.c0;

        Fp2 {
            c0: ac + Fp::new(NON_RESIDUE) * bd,
            c1: ad + bc,
        }
    }
}

impl Neg for Fp2 {
    type Output = Self;

    fn neg(self) -> Self {
        Fp2 {
            c0: -self.c0,
            c1: -self.c1,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add() {
        let a = Fp::new(123456789);
        let b = Fp::new(987654321);
        let c = a + b;
        assert_eq!(c.0, 1111111110);
    }

    #[test]
    fn test_mul() {
        let a = Fp::new(2);
        let b = Fp::new(3);
        let c = a * b;
        assert_eq!(c.0, 6);
    }

    #[test]
    fn test_inverse() {
        let a = Fp::new(12345);
        let a_inv = a.inverse();
        let should_be_one = a * a_inv;
        assert_eq!(should_be_one, Fp::ONE);
    }

    #[test]
    fn test_root_of_unity() {
        let omega = Fp::primitive_root_of_unity(4); // 16th root
        let omega_16 = omega.pow(16);
        assert_eq!(omega_16, Fp::ONE);
    }

    #[test]
    fn test_fp2_mul() {
        let a = Fp2::new(Fp::new(3), Fp::new(2)); // 3 + 2α
        let b = Fp2::new(Fp::new(1), Fp::new(4)); // 1 + 4α
        // (3 + 2α)(1 + 4α) = 3 + 12α + 2α + 8α^2 = 3 + 14α + 56 = 59 + 14α
        let c = a * b;
        assert_eq!(c.c0, Fp::new(3 + 8 * 7)); // 3 + 56 = 59
        assert_eq!(c.c1, Fp::new(14));
    }
}
